### ISSUE TITLE
Support Azure OpenAI service

### ISSUE THREAD

**Issue by LayZeeDK** [zurawiki/gptcommit#18]
Please add support for [Azure OpenAI service](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/overview?WT.mc_id=DT-MVP-5003831) that now has General Availability (GA), including the GPT-3 model.

**Comment by zurawiki** [comment]
Thanks for the suggestion @LayZeeDK. I'm interested by your request. What's motivating it?

I certainly agree that this project should be generalized to support different LLM APIs in the future. If we want to implement Azure, I'd be curious to learn:

- Is there some feature Azure offers over OpenAI? 
- Will the setup for the Azure API be any different, in terms of authentication or data format?

**Comment by LayZeeDK** [comment]
Azure OpenAI service supports enterprise-grade security, policies, compliance, and regional availability. Would you be interested in access?

**Comment by LayZeeDK** [comment]
API reference for Azure OpenAI service:
- [Completions and embeddings](https://learn.microsoft.com/en-us/azure/cognitive-services/openai/reference) [(Swagger)](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/cognitiveservices/data-plane/AzureOpenAI/inference/stable/2022-12-01/inference.json)

**Comment by zurawiki** [comment]
Yup this is definitely something I'd be interested in making happen! I tagged the issue as help wanted

**Comment by LayZeeDK** [comment]
# Request
In https://github.com/zurawiki/gptcommit/blob/6ee5bc587de6b7f2126d1cf073b8dc6e0b4d121c/src/openai.rs#L66-L71

Instead of `https://api.openai.com/v1/completions`, the endpoint URL would be in the format `https://YOUR_RESOURCE_NAME.openai.azure.com/openai/deployments/YOUR_DEPLOYMENT_NAME/embeddings?api-version=2022-12-01` so we must make the endpoint URL for completions configurable.

Instead of an `Authorization` header in the format `Bearer: <api_key>, we would have the header `api-key: <api_key>`

As of Azure OpenAI Service version `2022-12-01`, a completions request does not support a `model` parameter but currently we are not passing any.

# Response
The OpenAI API response object has a `usage` property but the Azure OpenAI Service API does not.

**Comment by zurawiki** [comment]
Currently blocked on async-openai supporting Azure: https://github.com/64bit/async-openai/issues/32

**Comment by Laisky** [comment]
You can try this project that converts Azure to OpenAI API.

<https://github.com/haibbo/cf-openai-azure-proxy/blob/main/README_en.md>

then edit the gptchat's config file(`~/.config/gptcommit/config.toml`)

![CleanShot 2023-05-16 at 10 05 50@2x](https://github.com/zurawiki/gptcommit/assets/4532436/05405484-3b19-4c70-8944-0ecf6744f916)

Note: `api_key` should be azure openai service's key, not openai's api key.
